### PR TITLE
Require minimum length for secret

### PR DESCRIPTION
**Pull request by wsargent**
## Fixes

Fixes https://github.com/playframework/playframework/issues/8554

## Purpose

Specifies a minimum length for the secret, below which the application will report an error.  This prevents low entropy strings from being used.

Note that this is a) set intentionally low so that no genuine random string is used and b) is hardcoded and can't be disabled.   

### Unfinished work

This needs to be mentioned in the migration notes, as it will prevent some users from updating (and will require logins as changing the salt will invalidate all sessions) and will also require changes for users who use a "prod" application in their test suites.

## Background Context

Salts aren't required to be incredibly strong, but they are required to be unique... and if you don't have a unique salt (i.e. you've been typing "password" in) then the application secret can be guessed and then the session data can be spoofed.

### CODE REVIEW

File: framework/src/play/src/main/scala/play/api/http/HttpConfiguration.scala [hunk 1]
```diff
@@ -251,6 +255,40 @@ object HttpConfiguration {
...
+        // https://crypto.stackexchange.com/a/34866 = 32 bytes (256 bits)
+        // https://security.stackexchange.com/a/11224 = (128 bits is more than enough)
+        // but if we have less than 8 bytes in production then it's not even 64 bits,
+        // which wasn't cool even in 2000 per RFC-2898 -- so set a lower bound of an
+        // 8 character secret, which is almost certainly base64 random entropy in any
+        // case, and is most probably a hardcoded text password.  The ENTIRE 8 character
+        // HMAC SHA256 space has been mapped out, so with a large enough dictionary,
+        // there will be a hit.
+        // https://tools.ietf.org/html/rfc2898#section-4.1
+        val message =
```
Comment (marcospereira) [inline framework/src/play/src/main/scala/play/api/http/HttpConfiguration.scala]: It could also be good to warn (or even fail?) in DEV mode if people are using a short secret key. It is possible many people are using the same key in DEV and PROD modes, so helping them to see the problem sooner would be good.

File: documentation/manual/releases/release27/migration27/Migration27.md [hunk 2]
```diff
@@ -397,3 +397,9 @@ CoordinatedShutdown.get(actorSystem).run(reason, runFromPhase);
 ## Change in self-signed HTTPS certificate
 
 It is now generated under `target/dev-mode/generated.keystore` instead of directly on the root folder.
+
+## Application Secret is checked for minimum length
+
+The [[application secret|ApplicationSecret]] configuration `play.http.secret.key` is checked for a minimum length in production.  If the key is fifteen characters or fewer, a warning will be logged.  If the key is eight characters or fewer, then an error is thrown and the configuration is invalid.  You can resolve this error by setting the secret to at least 32 bytes of competely random input, such as `cat /dev/urandom | base64` or by the application secret generator, using `playGenerateSecret` or `playUpdateSecret`.
```
Comment (marcospereira) [inline documentation/manual/releases/release27/migration27/Migration27.md]:
```suggestion
The [[application secret|ApplicationSecret]] configuration `play.http.secret.key` is checked for a minimum length in production.  If the key is fifteen characters or fewer, a warning will be logged.  If the key is eight characters or fewer, then an error is thrown and the configuration is invalid.  You can resolve this error by setting the secret to at least 32 bytes of completely random input, such as `head -c 32 /dev/urandom | base64` or by the application secret generator, using `playGenerateSecret` or `playUpdateSecret`.
```

File: documentation/manual/releases/release27/migration27/Migration27.md [hunk 3]
```diff
@@ -397,3 +397,9 @@ CoordinatedShutdown.get(actorSystem).run(reason, runFromPhase);
...
+
+The [[application secret|ApplicationSecret]] is used as the key for ensuring that a Play session cookie is valid, i.e. has been generated by the server as opposed to spoofed by an attacker.  However, the secret only specifies a string, and does not determine the amount of entropy in that string.  However, it is possible to put an upper bound on the amount of entropy in the secret simply by measuring how short it is: if the secret is eight characters long, that is at most 64 bits of entropy, which is insufficient by modern standards.
```
Comment (marcospereira) [inline documentation/manual/releases/release27/migration27/Migration27.md]:
```suggestion
The [[application secret|ApplicationSecret]] is used as the key for ensuring that a Play session cookie is valid, i.e. has been generated by the server as opposed to spoofed by an attacker.  However, the secret only specifies a string, and does not determine the amount of entropy in that string.  Anyhow, it is possible to put an upper bound on the amount of entropy in the secret simply by measuring how short it is: if the secret is eight characters long, that is at most 64 bits of entropy, which is insufficient by modern standards.
```

File: documentation/manual/releases/release27/migration27/Migration27.md (hunk 2 above)
Comment (marcospereira) [reply documentation/manual/releases/release27/migration27/Migration27.md]: Suggested to use `head -c 32` instead of `cat`.

File: documentation/manual/releases/release27/migration27/Migration27.md (hunk 3 above)
Comment (marcospereira) [reply documentation/manual/releases/release27/migration27/Migration27.md]: Just avoid two `However`s.

File: framework/src/play/src/main/scala/play/api/http/HttpConfiguration.scala (hunk 1 above)
Comment (wsargent) [reply framework/src/play/src/main/scala/play/api/http/HttpConfiguration.scala]: Sounds like a good idea, but only if you can turn off the warning for people who have already seen and heard it.  The default "changeme" config setting is a hell of a drug.

File: framework/src/play/src/main/scala/play/api/http/HttpConfiguration.scala [hunk 4]
```diff
@@ -239,8 +239,12 @@ object HttpConfiguration {
     )
   }
 
-  private def getSecretConfiguration(config: Configuration, environment: Environment): SecretConfiguration = {
+  @inline
+  private def secretEqualOrShorterThan(s: String, length: Int): Boolean = {
+    s.length <= 8
```
Comment (marcospereira) [inline framework/src/play/src/main/scala/play/api/http/HttpConfiguration.scala]:
```suggestion
    s.length <= length
```

File: framework/src/play/src/main/scala/play/api/http/HttpConfiguration.scala (hunk 1 above)
Comment (marcospereira) [reply framework/src/play/src/main/scala/play/api/http/HttpConfiguration.scala]: Well, you can turn then off by configuring the log for `play.api.HttpConfiguration`. Wouldn't that be enough?

**Comment by marcospereira** [comment]
@wsargent you may want to do a small review here since I've added some commits. 😄

File: framework/src/play-integration-test/src/test/scala/play/it/api/SecretConfigurationParserSpec.scala [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+      }
+
+      "log a warning when secret length is smaller than 8 chars" in {
+        val (secret, events) = parseSecret(mode = Mode.Dev)("play.http.secret.key" -> "1234567")
+        events.map(_.getFormattedMessage).find(_.contains("The application secret is too short and does not have the recommended amount of entropy")) must beSome
+        secret must beSome
+      }
+
+      "log a warning when secret length is smaller then 15 chars" in {
+        val (secret, events) = parseSecret(mode = Mode.Dev)("play.http.secret.key" -> "1234567890123")
```
Comment (ignasi35) [inline framework/src/play-integration-test/src/test/scala/play/it/api/SecretConfigurationParserSpec.scala]: `secret` is 13 chars long. Should be 14 to test the edge case.
```suggestion
        val (secret, events) = parseSecret(mode = Mode.Dev)("play.http.secret.key" -> "12345678901234")
```

File: framework/src/play/src/main/scala/play/api/http/HttpConfiguration.scala [hunk 6]
```diff
@@ -251,6 +261,58 @@ object HttpConfiguration {
...
+      case Some(s) if secretEqualOrShorterThan(s, 15) && !s.equals("changeme") && s.trim.nonEmpty && environment.mode == Mode.Dev =>
+        val message =
+          """
+            |Your secret key is very short, and may be vulnerable to dictionary attacks.  Your application may not be secure.
+            |The application secret should ideally be 32 bytes of completely random input, encoded in base64. While the application
+            |will be able to start in production mode, you will also see a warning when it is starting.
+            |To set the application secret, please read http://playframework.com/documentation/latest/ApplicationSecret
+          """.stripMargin
+        logger.warn(message)
+        s
```
Comment (ignasi35) [inline framework/src/play/src/main/scala/play/api/http/HttpConfiguration.scala]: `8` and `15` are magic numbers used in two places each, I think we could extract the values.

Comment (ignasi35) [reply framework/src/play/src/main/scala/play/api/http/HttpConfiguration.scala]: Then, I'd move the code comments wrt length to the values declaration (moving them away from the pattern matching).

File: framework/src/play/src/test/scala/play/api/http/HttpConfigurationSpec.scala [hunk 7]
```diff
@@ -48,8 +48,8 @@ class HttpConfigurationSpec extends Specification {
         "play.http.flash.jwt.clockSkew" -> "30s",
         "play.http.flash.jwt.dataClaim" -> "data",
         "play.http.fileMimeTypes" -> "foo=text/foo",
-        "play.http.secret.key" -> "mysecret",
-        "play.http.secret.provider" -> "HmacSHA1"
+        "play.http.secret.key" -> "ad31779d4ee49d5ad5162bf1429c32e2e9933f3b",
+        "play.http.secret.provider" -> "HmacSHA256"
```
Comment (ignasi35) [inline framework/src/play/src/test/scala/play/api/http/HttpConfigurationSpec.scala]: How will this default change affect a rolling upgrade?

Will cookies produced in `node-1` (a node still not updated using HmacSHA1) be readable from `node-2` (which is already updated and expects cookies signed with `HmacSHA256`? (or vice-versa)

File: framework/src/play-integration-test/src/test/scala/play/it/api/SecretConfigurationParserSpec.scala [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+        events.map(_.getFormattedMessage).find(_.contains("Your secret key is very short, and may be vulnerable to dictionary attacks.  Your application may not be secure")) must beSome
+        secret must beSome
+      }
+
+      "return the value without warnings when it is configured respecting the requirements" in {
+        val (secret, events) = parseSecret(mode = Mode.Dev)("play.http.secret.key" -> "12345678901234567890")
```
Comment (ignasi35) [inline framework/src/play-integration-test/src/test/scala/play/it/api/SecretConfigurationParserSpec.scala]: Make the secret respect the requirements using an edge case: 16 chars?

File: framework/src/play-integration-test/src/test/scala/play/it/api/SecretConfigurationParserSpec.scala [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+        secret must beNone
+      }
+
+      "fail when value length is smaller than 8 chars" in {
+        val (secret, _) = parseSecret(mode = Mode.Prod)("play.http.secret.key" -> "1234567")
+        secret must beNone
+      }
+
+      "log a warning when value length is smaller than 15 chars" in {
+        val (secret, events) = parseSecret(mode = Mode.Prod)("play.http.secret.key" -> "1234567890123")
```
Comment (ignasi35) [inline framework/src/play-integration-test/src/test/scala/play/it/api/SecretConfigurationParserSpec.scala]:
```suggestion
        val (secret, events) = parseSecret(mode = Mode.Prod)("play.http.secret.key" -> "123456789012345")
```

I noticed on the code below the actual conditions are not `smaller than` `8` or `15` but `smaller than or equal` so the wording of the test case and the value of the secret need reviewing.

File: framework/src/play/src/main/scala/play/api/http/HttpConfiguration.scala (hunk 6 above)
Comment (wsargent) [reply framework/src/play/src/main/scala/play/api/http/HttpConfiguration.scala]: Fixed

File: framework/src/play/src/test/scala/play/api/http/HttpConfigurationSpec.scala (hunk 7 above)
Comment (wsargent) [reply framework/src/play/src/test/scala/play/api/http/HttpConfigurationSpec.scala]: This isn't a default change, it's a spec fix.  The cookies are generated by JWT now, which is a distinct mechanism https://www.playframework.com/documentation/2.6.x/Migration26#jwt-support

The `play.http.secret.provider` is really and truly out of date -- it's the JCA provider, which is going to be the out of the box one and would be "SunJCE" if you defined it explicitly... so both settings are wrong there.  

https://github.com/playframework/playframework/blob/master/framework/src/play/src/main/resources/reference.conf#L232

File: framework/src/play/src/main/scala/play/api/http/HttpConfiguration.scala (hunk 1 above)
Comment (wsargent) [reply framework/src/play/src/main/scala/play/api/http/HttpConfiguration.scala]: That's fair!

File: framework/src/play-integration-test/src/test/scala/play/it/api/SecretConfigurationParserSpec.scala (hunk 9 above)
Comment (wsargent) [reply framework/src/play-integration-test/src/test/scala/play/it/api/SecretConfigurationParserSpec.scala]: Fixed

**Comment by wsargent** [comment]
@ignasi35 should be good now

File: framework/src/play-integration-test/src/test/scala/play/it/api/SecretConfigurationParserSpec.scala (hunk 8 above)
Comment (wsargent) [reply framework/src/play-integration-test/src/test/scala/play/it/api/SecretConfigurationParserSpec.scala]: Fixed
